### PR TITLE
Update configs

### PR DESCRIPTION
**Pull request by Goldziher**
This PR does several things:

1. I updated all packages. This includes the replacement of vue-popperjs with popper.js v2 (we were actullly using the popper.js package from the vue-popperjs, rather than the component, and this was an old package), as well as an update to all existing packages.

2. I update the build configs, so when kotti is built it uses the latest settings from vue/cli 4, does polyfilling correctly and optimizes css + fixes as we do in yoda (better browser compatibility).

3. I updated the nuxt configs to support core-js 3, which we use in the vue cli building. For some reason nuxt is slower to adopt this. 

4. I replaced the weird fork of normalize.css we had in assets, with the updated csstools/normalize package, which is part of the post-css ecosystem. This is the same thing as we had before, only five major versions more updated.

5. I added eslint, stylelint, babel and post css configs, and update pre-commit hooks to use these correctly. 

6. I linted the entire project and fixed multiple bugs. There are still some remianing but these will require attention later on. 

7. I updated the package.json scripts with some useful commands. 

Result - we have better packaging and building for the kotti npm package. the github pages / netlify deploys work as expected. Kotti is now aligned with yoda, so we will not have issues when embedding it. This also paves the way to begin migration to TS based vue.

### CODE REVIEW

**Comment by FlorianWendelborn** [comment]
Looking at `deploy/netlify` `Details`, it seems that "the weird fork of normalize" is actually not doing the same thing. There’s a white border around the entire document now: https://deploy-preview-140--kotti.netlify.com/

**Comment by Goldziher** [comment]
@FlorianWendelborn - I don't think this is the root cause of this. The styles are not used by the nuxt build, they are extracted when kotti is built as a package. I'm not 100% certain though so I will check. 

As for the "weirdness" - junyu manually forked a now deprecated version of normalize.css. Google normalize.css and you'll find v8 of the original that he forked (he forked v5). This has since been moved into a repository belonging to csstools, and the css document there is primarily the same (you can visually compare the two or create a diff locally to see) .

File: .prettierrc.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+module.exports = {
+	quoteProps: 'consistent',
```
Comment (FlorianWendelborn) [inline .prettierrc.js]: I’m voting in favor of removing this since this will cause bigger diffs for no benefit that’s known to me. Imagine someone adds a `background-color` to an already existing style object. Now, the entire object gets re-formatted.

File: packages/kotti-table/src/logic/order.js [hunk 2]
```diff
@@ -24,6 +24,7 @@ export const mutations = {
 			col.order = index
 			col.index = index
 		})
+		//FIXME: there is an issue in the line below
```
Comment (FlorianWendelborn) [inline packages/kotti-table/src/logic/order.js]: Please describe the issue

File: package.json [hunk 3]
```diff
@@ -41,79 +39,100 @@
...
-		"trailingComma": "all",
-		"useTabs": true
-	},
 	"scripts": {
 		"build:gh-pages": "DEPLOY_ENV=GH_PAGES nuxt generate",
 		"build:kotti": "vue-cli-service build --target lib --name KottiUI packages/index.js",
 		"deploy:docs": "./github-deploy.sh",
 		"dev:docs": "nuxt",
-		"precommit": "lint-staged",
-		"prepublishOnly": "yarn run build:kotti"
```
Comment (FlorianWendelborn) [inline package.json]: Why was this removed?

Comment (Goldziher) [reply package.json]: prepublishOnly is the same as yarn build:kotti, its just an alias and it was not used anywhere. precommit, well runs the hook. If you are using this I'll return it, but I'd think that for the rare use of this - yarn pre-commit in the terminal would do.

File: .eslintignore [hunk 4]
```diff
@@ -0,0 +1,4 @@
+dist/*
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply .eslintignore]: fixed

File: .eslintrc.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+		'plugin:sonarjs/recommended',
+		'plugin:prettier/recommended',
+		'prettier/vue',
+	],
+	rules: {
+		'no-magic-numbers': [
+			'warn',
+			{ ignoreArrayIndexes: true, ignore: [0, 1, 2, -1, 10] },
+		],
+		'prettier/prettier':  'error',
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply .eslintrc.js]: fixed

File: .eslintrc.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+		'import/max-dependencies': 'off',
+		'import/order': 'off',
+		'no-param-reassign':  'warn',
+		'linebreak-style': ['error', 'unix'],
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply .eslintrc.js]: indeed

File: .eslintrc.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+		'no-console':  'error',
+		'no-debugger': 'error',
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply .eslintrc.js]: fixed

File: .prettierrc.js (hunk 1 above)
Comment (Goldziher) [reply .prettierrc.js]: i aligned with yoda. Not attached to this at all, but IMO we should align all repos to use similar settings (i don't care which) @mvangent?

File: packages/kotti-table/src/logic/order.js (hunk 2 above)
Comment (Goldziher) [reply packages/kotti-table/src/logic/order.js]: fixed it - was a function that returned nothing --> @carsoli

File: .eslintrc.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+		'prettier/vue',
+	],
+	rules: {
+		'no-magic-numbers': [
+			'warn',
+			{ ignoreArrayIndexes: true, ignore: [0, 1, 2, -1, 10] },
+		],
+		'prettier/prettier':  'error',
+		'import/max-dependencies': 'off',
+		'import/order': 'off',
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply .eslintrc.js]: removed

File: .eslintrc.js [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+		'eslint:recommended',
+		'plugin:vue/recommended',
+		'plugin:sonarjs/recommended',
+		'plugin:prettier/recommended',
+		'prettier/vue',
+	],
+	rules: {
+		'no-magic-numbers': [
+			'warn',
+			{ ignoreArrayIndexes: true, ignore: [0, 1, 2, -1, 10] },
```
Comment (carsoli) [inline .eslintrc.js]: why the 10 though?

File: .eslintrc.js [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+		'import/order': 'off',
+		'no-param-reassign':  'warn',
+		'linebreak-style': ['error', 'unix'],
+		'no-console':  'error',
+		'no-debugger': 'error',
+		'vue/attribute-hyphenation': ['error', 'never'],
+		'vue/require-default-prop': 'off',
+		'vue/require-component-is':  'warn',
+		'vue/html-indent': 'off',
+		'sonarjs/no-duplicate-string':  'warn',
```
Comment (carsoli) [inline .eslintrc.js]: if we are not using an eslint-plugin for sonarjs 
i suggest we add more rules
for reference: https://github.com/SonarSource/eslint-plugin-sonarjs

File: .prettierrc.js (hunk 1 above)
Comment (FlorianWendelborn) [reply .prettierrc.js]: I agree that repos should be aligned, so also voting in favor of removing it in Yoda.

File: package.json (hunk 3 above)
Comment (FlorianWendelborn) [reply package.json]: `prepublishOnly` automatically runs before you publish a package, making sure it’s actually built with the latest version. It’s one of the npm lifecycle hooks

File: .eslintrc.js (hunk 9 above)
Comment (FlorianWendelborn) [reply .eslintrc.js]: 10 is often used for base10 conversions and is pretty easy to understand in most cases

File: .eslintrc.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (carsoli) [reply .eslintrc.js]: also, have you pushed the commit the removes this line? because i still see it in the overall diff @Goldziher

File: .prettierrc.js (hunk 1 above)
Comment (carsoli) [reply .prettierrc.js]: @FlorianWendelborn in that case which option would we consider: 'as needed', maybe?

File: .eslintrc.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (Goldziher) [reply .eslintrc.js]: https://github.com/3YOURMIND/kotti/pull/140/commits/158b37b251bc318840436309dbf74da8315e954f

[the remark this answers is not in the document]
Comment (Goldziher) [reply .eslintrc.js]: sorry, this https://github.com/3YOURMIND/kotti/pull/140/commits/158b37b251bc318840436309dbf74da8315e954f

File: package.json [hunk 11]
```diff
@@ -41,79 +39,100 @@
...
 		"dist",
 		"packages/kotti-style/_variables.scss"
 	],
 	"license": "MIT",
+	"husky": {
+		"hooks": {
+			"pre-commit": "lint-staged"
+		}
+	},
 	"lint-staged": {
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply package.json]: I read the docs there. I'm against this one for two reaons - 1. we use lint staged in yoda, because of 2.  we use stylelint and eslint as well, which is the reason for option 1 in the document you linked.

File: .prettierrc.js (hunk 1 above)
Comment (Goldziher) [reply .prettierrc.js]: well, feel free to push to this PR with whatever you guys decide - but we will then need to reformat yoda. Might be a good time to do a full linting as well, and get the big diff over with.

File: package.json (hunk 3 above)
Comment (Goldziher) [reply package.json]: i see, cool. i'll return this.

File: .eslintrc.js (hunk 10 above)
Comment (Goldziher) [reply .eslintrc.js]: i think we use all of the recommended ones - the we extend the plugin. its just overwriting this from error to warning.

File: package.json (hunk 3 above)
Comment (Goldziher) [reply package.json]: returned it

File: .eslintrc.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (Goldziher) [reply .eslintrc.js]: i don't really understand the github UI. anyhow, yes this doesn't exist in the branch anymore, i dunno how to see the current diff though.

File: package.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (Goldziher) [reply package.json]: we use stylelint and eslint as well as part of lint staged. does precise-commits support this?

File: .github/ISSUE_TEMPLATE/bug_report.md [hunk 12]
```diff
@@ -1,17 +1,16 @@
...
 name: Bug report
 about: Create a report to help us improve
 title: ''
 labels: ''
 assignees: ''
-
----
-
-**Describe the bug**
+---**Describe the bug**
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply .github/ISSUE_TEMPLATE/bug_report.md]: resolved

File: .github/ISSUE_TEMPLATE/feature_request.md [hunk 13]
```diff
@@ -1,13 +1,11 @@
...
 name: Feature request
 about: Suggest an idea for this project
 title: ''
 labels: ''
 assignees: ''
-
----
-
-**Is your feature request related to a problem? Please describe.**
+---**Is your feature request related to a problem? Please describe.**
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply .github/ISSUE_TEMPLATE/feature_request.md]: fixed

File: .nvmrc [hunk 14]
```diff
@@ -0,0 +1 @@
+10
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply .nvmrc]: fixed

File: babel.config.js [hunk 15]
```diff
@@ -0,0 +1,10 @@
+module.exports = (api) => {
+	api.cache(false)
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply babel.config.js]: fixed

File: nuxt.config.js [hunk 16]
```diff
@@ -40,6 +39,20 @@ design decisions. The Kotti design system has two main parts: foundation and com
 	 ** Build configuration
 	 */
 	build: {
+		babel: {
+			presets({ isServer }) {
+				return [
+					[
+						require.resolve('@nuxt/babel-preset-app'),
+						// require.resolve('@nuxt/babel-preset-app-edge'), // For nuxt-edge users
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply nuxt.config.js]: fixed

File: package.json [hunk 17]
```diff
@@ -41,79 +39,100 @@
...
+		"vue-clickaway": "^2.2.2"
 	},
 	"description": "Kotti Vue Component UI",
 	"devDependencies": {
-		"@nuxtjs/axios": "^5.4.0",
-		"@nuxtjs/markdownit": "^1.2.0",
+		"@babel/core": "^7.8.4",
+		"@babel/plugin-proposal-nullish-coalescing-operator": "^7.8.3",
+		"@babel/plugin-proposal-optional-chaining": "^7.8.3",
+		"@csstools/normalize.css": "^10.1.0",
```
[the remark this answers is not in the document]
Comment (carsoli) [reply package.json]: I don't get it? You removed a normalization file and added this as a dependency, no? @Goldziher

File: packages/kotti-table/src/logic/order.js (hunk 2 above)
Comment (carsoli) [reply packages/kotti-table/src/logic/order.js]: i don't see the fix but it's a missing return statement in the exported function, not the getter
should work on issues of the table again soon (next tech day probably) so we can report this as an issue now since it's out of this scope @Goldziher

File: stylelint.config.js [hunk 18]
```diff
@@ -0,0 +1,21 @@
+// eslint-disable-next-line
```
Comment (carsoli) [inline stylelint.config.js]: why are we disabling linter here @Goldziher

File: package.json [hunk 19]
```diff
@@ -6,10 +6,8 @@
 		]
 	},
 	"browserslist": [
-		"> 1%",
-		"last 2 versions",
-		"ie 11",
-		"not ie <= 8"
+		"defaults",
```
Comment (carsoli) [inline package.json]: is this the configuration for yoda as well?

File: nuxt.config.js [hunk 20]
```diff
@@ -40,6 +39,20 @@ design decisions. The Kotti design system has two main parts: foundation and com
 	 ** Build configuration
 	 */
 	build: {
+		babel: {
+			presets({ isServer }) {
+				return [
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply nuxt.config.js]: i copied the example from the docs 1:1 (except a commented out example). I'd guess each sub array excepts an optional config object or something.

File: package.json [hunk 21]
```diff
@@ -41,79 +39,100 @@
...
-		"useTabs": true
-	},
 	"scripts": {
 		"build:gh-pages": "DEPLOY_ENV=GH_PAGES nuxt generate",
 		"build:kotti": "vue-cli-service build --target lib --name KottiUI packages/index.js",
 		"deploy:docs": "./github-deploy.sh",
 		"dev:docs": "nuxt",
-		"precommit": "lint-staged",
-		"prepublishOnly": "yarn run build:kotti"
+		"lint:script": "eslint --fix '**/*.{js,vue}'",
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply package.json]: the terminal script will fail if no files match the extension in the argument - so we need to have some of those first :)

File: package.json [hunk 22]
```diff
@@ -41,79 +39,100 @@
...
-		"@3yourmind/yoco": "~1.1.0",
-		"color": "~3.1.0",
-		"dayjs": "~1.8.8",
+		"@3yourmind/yoco": "^1.3.1",
+		"@popperjs/core": "^2.0.5",
+		"color": "^3.1.2",
+		"dayjs": "^1.8.20",
 		"deep-eql": "^4.0.0",
-		"eslint": "^6.8.0",
-		"eslint-plugin-prettier": "^3.1.2",
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply package.json]: we're not - i just sorted imports alphabetically.

File: jsconfig.json [hunk 23]
```diff
@@ -0,0 +1,7 @@
+{
+	"compilerOptions": {
+		"allowSyntheticDefaultImports": true,
+		"baseUrl": "."
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply jsconfig.json]: it depends on the scope in which you open the folder - jsconfig.json is used by vscode, its a microsoft thingy. So if you open a folder from a different scope in the workspace this is supposedly used for resolving paths and the like. Saying that, it may well be redundant.

File: package.json [hunk 24]
```diff
@@ -41,79 +39,100 @@
...
-		"*.{js,json,css,scss,md,vue}": [
-			"prettier --write",
-			"git add"
+		"*.{json,md}": [
+			"prettier --write"
+		],
+		"*.{css,scss,vue}": [
+			"stylelint --fix"
+		],
+		"*.{js,vue,ts,tsx}": [
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply package.json]: sure, but we are not likely to use it - since we will probably do tsx, if at all (not much point with vue files working with ts, either).

File: packages/kotti-user-menu/src/UserMenu.vue [hunk 25]
```diff
@@ -79,7 +79,7 @@ export default {
 		sections: {
 			type: Array,
 			required: true,
-			validator: sections => sections.every(sectionIsValid),
+			validator: (sections) => sections.every(sectionIsValid),
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply packages/kotti-user-menu/src/UserMenu.vue]: this is prettier, which is applied to everything.

File: package.json (hunk 19 above)
Comment (Goldziher) [reply package.json]: yup. this is some old set they have there - basically we need pollyfilling for ie11, and then for the common defaults, which cover all major browsers plus a bunch we don't use often but are quite common (opera etc.).

File: stylelint.config.js (hunk 18 above)
Comment (Goldziher) [reply stylelint.config.js]: i'll clean this (if its actually, wrong, will check).

File: package.json (hunk 24 above)
[the remark this answers is not in the document]
Comment (Goldziher) [reply package.json]: i'll add this so its complete :)

**Comment by Goldziher** [comment]
ok, @FlorianWendelborn @mvangent @carsoli @tomquinonero - i've created a test setup that seems to work for yoda. I'll make this into a mock PR, and add you all as reviewers. You can test how this looks in yoda and we can make adjustments if required - to me it seemed fine after i changed a bit the postcss setup.

File: .prettierrc.js (hunk 1 above)
Comment (mvangent) [reply .prettierrc.js]: I agree as-needed (the default) is preferred over consistent.

**Comment by tomquinonero** [comment]
Did not review the entire code, but @Goldziher showed it to me this morning and I agree on the principle. Approved

File: .prettierrc.js (hunk 1 above)
Comment (Goldziher) [reply .prettierrc.js]: as-needed it is then

File: packages/kotti-datepicker/src/DatePicker.vue [hunk 26]
```diff
@@ -61,7 +61,7 @@ export default {
 		monthsTranslations: {
 			type: Array,
 			default: () =>
-				range(0, 11).map(month =>
+				range(0, 11).map((month) =>
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply packages/kotti-datepicker/src/DatePicker.vue]: nope, but these are warnings. im . not going to fix all of then - there are 30-40, you can run "yarn check:script" to see them

File: packages/kotti-inline-edit/src/InlineEdit.vue [hunk 27]
```diff
@@ -30,8 +30,8 @@ import escape from 'lodash/escape'
 import KtButton from '../../kotti-button'
 import KtButtonGroup from '../../kotti-button-group'
 
-const DEFAULT_POST_PARSER = _ => _
-const newLineParser = t => t.replace(/\n/gm, '<br/>')
+const DEFAULT_POST_PARSER = (_) => _
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply packages/kotti-inline-edit/src/InlineEdit.vue]: they are

File: packages/kotti-inline-edit/index.js [hunk 28]
```diff
@@ -1,7 +1,7 @@
 import KtInlineEdit from './src/InlineEdit'
 import escape from 'lodash/escape'
 
-export const defaultParser = message => escape(message)
+export const defaultParser = (message) => escape(message)
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply packages/kotti-inline-edit/index.js]: prettier
